File: app/src/main/java/com/bitflaker/lucidsourcekit/alarms/OutsideSlider.java
package com.bitflaker.lucidsourcekit.alarms;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.LinearInterpolator;

import androidx.core.content.res.ResourcesCompat;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.general.Tools;

public class OutsideSlider extends View {
    private final Paint indicatorPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
    private final Paint dataLinePaint = new Paint();
    private final Paint dataLinePaintBackground = new Paint();
    private Bitmap leftIcon, rightIcon, leftArrow, rightArrow;
    private OnLeftSideSelected mLeftSideSelectedListener;
    private OnRightSideSelected mRightSideSelectedListener;
    private OnFadedAway mFadedAwayListener;
    private boolean hasToSwipe = false;
    private boolean isHoldSwiping = false;
    private float pressPointOffset = 0;
    private float buttonSelectorRadius;
    private float xButtonSelector = -1;
    private float sidePadding = Tools.dpToPx(getContext(), 32);
    private float currentArrowMargin;
    private int primColor = Tools.getAttrColor(R.attr.primaryTextColor, getContext().getTheme());
    private ValueAnimator valAnim;
    private int indicatorMoveTo = Tools.dpToPx(getContext(), 50);
    private boolean isSwiping = false;

    public OutsideSlider(Context context){
        super(context);
        setup();
    }

    public OutsideSlider(Context context, AttributeSet attributeSet){
        super(context, attributeSet);
        setup();
    }

    private void setup() {
        setLayerType(View.LAYER_TYPE_HARDWARE, null);
        primColor = Tools.getAttrColor(R.attr.primaryTextColor, getContext().getTheme());
        dataLinePaintBackground.setColor(Tools.getAttrColor(R.attr.backgroundColor, getContext().getTheme()));
        dataLinePaintBackground.setStrokeCap(Paint.Cap.ROUND);
        dataLinePaintBackground.setAntiAlias(true);
        dataLinePaintBackground.setStrokeWidth(0);
        dataLinePaint.setColor(Tools.getAttrColor(R.attr.primaryTextColor, getContext().getTheme()));
        dataLinePaint.setStrokeWidth(Tools.dpToPx(getContext(), 2));
        dataLinePaint.setStyle(Paint.Style.STROKE);
        dataLinePaint.setAntiAlias(true);
        indicatorPaint.setColor(primColor);
        indicatorPaint.setAntiAlias(true);
        buttonSelectorRadius = Tools.dpToPx(getContext(), 24);

        this.leftIcon = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_baseline_check_24, getContext().getTheme()), primColor, Tools.dpToPx(getContext(), 32));
        this.rightIcon = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_baseline_snooze_24, getContext().getTheme()), primColor, Tools.dpToPx(getContext(), 32));
        this.leftArrow = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_round_keyboard_double_arrow_left_24, getContext().getTheme()), Tools.dpToPx(getContext(), 24));
        this.rightArrow = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_round_keyboard_double_arrow_right_24, getContext().getTheme()), Tools.dpToPx(getContext(), 24));
    }

    public void setData(Drawable leftIcon, Drawable rightIcon) {
        this.leftIcon = Tools.drawableToBitmap(leftIcon, primColor, Tools.dpToPx(getContext(), 32));
        this.rightIcon = Tools.drawableToBitmap(rightIcon, primColor, Tools.dpToPx(getContext(), 32));
        this.leftArrow = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_round_keyboard_double_arrow_left_24, getContext().getTheme()), Tools.dpToPx(getContext(), 24));
        this.rightArrow = Tools.drawableToBitmap(ResourcesCompat.getDrawable(getResources(), R.drawable.ic_round_keyboard_double_arrow_right_24, getContext().getTheme()), Tools.dpToPx(getContext(), 24));
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if(xButtonSelector == -1){
            xButtonSelector = getWidth() / 2.0f;
            valAnim = ValueAnimator.ofInt(0, indicatorMoveTo);
            valAnim.setDuration(1300);
            valAnim.setInterpolator(new LinearInterpolator());
            valAnim.addUpdateListener((valueAnimator) -> {
                currentArrowMargin = (int) valueAnimator.getAnimatedValue();
                invalidate();
            });
            valAnim.setRepeatCount(-1);
            valAnim.setRepeatMode(ValueAnimator.RESTART);
            valAnim.start();
        }
        indicatorPaint.setColor(Tools.manipulateAlpha(primColor, getAlphaValue()));
        canvas.drawBitmap(leftArrow, getWidth()/2.0f - sidePadding / 2.0f - leftArrow.getWidth() - currentArrowMargin, (getHeight() - leftArrow.getHeight())/2.0f, indicatorPaint);
        canvas.drawBitmap(rightArrow, getWidth()/2.0f + sidePadding / 2.0f + currentArrowMargin, (getHeight() - leftArrow.getHeight())/2.0f, indicatorPaint);
        canvas.drawBitmap(leftIcon, sidePadding, (getHeight() - leftIcon.getHeight())/2.0f, dataLinePaint);
        canvas.drawBitmap(rightIcon, getWidth() - rightIcon.getWidth() - sidePadding, (getHeight() - rightIcon.getHeight())/2.0f, dataLinePaint);
        canvas.drawCircle(xButtonSelector, getHeight()/2.0f, buttonSelectorRadius, dataLinePaint);
    }

    private float getAlphaValue() {
        float pos = currentArrowMargin / (float)indicatorMoveTo;
        float finalVal = 0.0f;
        if(hasToSwipe && !isHoldSwiping || !hasToSwipe && !isSwiping){
            finalVal = (float) Math.max(Math.min(7.48164*Math.pow(pos, 3)-15.36474*Math.pow(pos, 2)+7.88309*pos, 1), 0);
        }
        return finalVal;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch(event.getAction()){
            case MotionEvent.ACTION_DOWN:
                checkIfIsSwiping(event);
                isSwiping = true;
            case MotionEvent.ACTION_MOVE:
                setButtonPosition(event);
                break;
            case MotionEvent.ACTION_UP:
                setButtonPosition(event);
                isHoldSwiping = false;
                isSwiping = false;
                switch(getHoverOverSide()) {
                    case NONE:
                        resetButtonPos();
                        break;
                    case LEFT:
                        isHoldSwiping = true; // just not to get the arrows again
                        isSwiping = true; // just not to get the arrows again
                        if(mLeftSideSelectedListener != null){
                            mLeftSideSelectedListener.onEvent();
                        }
                        fadeOut();
                        break;
                    case RIGHT:
                        isHoldSwiping = true; // just not to get the arrows again
                        isSwiping = true; // just not to get the arrows again
                        if(mRightSideSelectedListener != null){
                            mRightSideSelectedListener.onEvent();
                        }
                        fadeOut();
                        break;
                }
                break;
        }
        invalidate();
        return true;
    }

    private void fadeOut() {
        ValueAnimator opacityAnim = ValueAnimator.ofFloat(1, 0);
        opacityAnim.setDuration(300);
        opacityAnim.setInterpolator(new LinearInterpolator());
        opacityAnim.addUpdateListener((valueAnimator) -> {
            setAlpha((float)valueAnimator.getAnimatedValue());
            if((float)valueAnimator.getAnimatedValue() == 0.0f){
                if(mFadedAwayListener != null){
                    mFadedAwayListener.onEvent();
                }
            }
        });
        opacityAnim.start();
    }

    private void checkIfIsSwiping(MotionEvent event) {
        boolean isOnX = event.getX() >= xButtonSelector - buttonSelectorRadius && event.getX() <= xButtonSelector + buttonSelectorRadius;
        float ySpace = (getHeight() - 2*buttonSelectorRadius)/2.0f;
        boolean isOnY = event.getY() >= ySpace && event.getY() <= getHeight() - ySpace;
        isHoldSwiping = isOnX && isOnY;
    }

    private void resetButtonPos() {
        xButtonSelector = getWidth() / 2.0f;
        valAnim.cancel();
        valAnim.start();
    }

    private void setButtonPosition(MotionEvent event) {
        if(!hasToSwipe || isHoldSwiping){
            xButtonSelector = Math.min(Math.max(event.getX(), sidePadding + leftIcon.getWidth()/2.0f), getWidth()-sidePadding-rightIcon.getWidth()/2.0f);
        }
    }

    private HoveringOverSide getHoverOverSide() {
        if(xButtonSelector <= sidePadding + leftIcon.getWidth()){
            return HoveringOverSide.LEFT;
        }
        else if(xButtonSelector >= getWidth() - sidePadding - rightIcon.getWidth()){
            return HoveringOverSide.RIGHT;
        }
        return HoveringOverSide.NONE;
    }

//    public static Bitmap drawableToBitmap (Drawable drawable, int size) {
//        if (drawable instanceof BitmapDrawable) {
//            return ((BitmapDrawable)drawable).getBitmap();
//        }
//
//        Bitmap bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
//        Canvas canvas = new Canvas(bitmap);
//        drawable.setBounds(0, 0, size, size);
//        drawable.draw(canvas);
//
//        return bitmap;
//    }

    public boolean isHasToSwipe() {
        return hasToSwipe;
    }

    public void setHasToSwipe(boolean hasToSwipe) {
        this.hasToSwipe = hasToSwipe;
    }

    enum HoveringOverSide {
        NONE,
        LEFT,
        RIGHT
    }

    public interface OnLeftSideSelected {
        void onEvent();
    }

    public void setOnLeftSideSelectedListener(OnLeftSideSelected eventListener) {
        mLeftSideSelectedListener = eventListener;
    }

    public interface OnRightSideSelected {
        void onEvent();
    }

    public void setOnRightSideSelectedListener(OnRightSideSelected eventListener) {
        mRightSideSelectedListener = eventListener;
    }

    public interface OnFadedAway {
        void onEvent();
    }

    public void setOnFadedAwayListener(OnFadedAway eventListener) {
        mFadedAwayListener = eventListener;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/views/OutsideSlider.kt
package com.bitflaker.lucidsourcekit.views

import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.animation.LinearInterpolator
import androidx.core.graphics.minus
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.utils.Tools.manipulateAlpha
import com.bitflaker.lucidsourcekit.utils.attrColor
import com.bitflaker.lucidsourcekit.utils.dpToPx
import com.bitflaker.lucidsourcekit.utils.resolveDrawableBitmap
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow

class OutsideSlider @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {
    // Pre-configured paints
    private var primColor = context.attrColor(R.attr.primaryTextColor)
    private val indicatorPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG).apply {
        color = primColor
    }
    private val dataLinePaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG).apply {
        color = context.attrColor(R.attr.primaryTextColor)
        style = Paint.Style.STROKE
        strokeWidth = 2f.dpToPx
    }

    // Configurable view specific fields
    var leftIcon = context.resolveDrawableBitmap(R.drawable.ic_baseline_check_24, primColor, 32.dpToPx)!!
    var rightIcon = context.resolveDrawableBitmap(R.drawable.ic_baseline_snooze_24, primColor, 32.dpToPx)!!
    var leftArrow = context.resolveDrawableBitmap(R.drawable.ic_round_keyboard_double_arrow_left_24, 24.dpToPx)!!
    var rightArrow = context.resolveDrawableBitmap(R.drawable.ic_round_keyboard_double_arrow_right_24, 24.dpToPx)!!
    var requiresSwipeMotion: Boolean = false
    var buttonSelectorRadius = 24f.dpToPx
    var sidePadding = 32f.dpToPx
    var indicatorMoveTo = 50.dpToPx

    // Event listeners
    var onLeftSideSelectedListener: (() -> Unit)? = null
    var onRightSideSelectedListener: (() -> Unit)? = null
    var onFadedAwayListener: (() -> Unit)? = null

    // Internal values
    private var dragButtonPosition = PointF()
    private var currentArrowMargin = 0f
    private var isHoldSwiping = false
    private var isSwiping = false
    private var arrowAnimation = ValueAnimator.ofInt(0, indicatorMoveTo).apply {
        duration = 1300
        interpolator = LinearInterpolator()
        repeatCount = -1
        repeatMode = ValueAnimator.RESTART
        addUpdateListener {
            currentArrowMargin = (it.getAnimatedValue() as Int).toFloat()
            invalidate()
        }
    }

    init {
        setLayerType(LAYER_TYPE_HARDWARE, null)
        arrowAnimation.start()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Set the arrow icon alpha to the appropriate value of the animation
        indicatorPaint.color = manipulateAlpha(primColor, getCurrentAnimatedAlpha())

        // Draw the animated left and right arrows
        canvas.drawBitmap(
            leftArrow,
            (width - sidePadding) / 2.0f - leftArrow.width - currentArrowMargin,
            (height - leftArrow.height) / 2.0f,
            indicatorPaint
        )
        canvas.drawBitmap(
            rightArrow,
            (width + sidePadding) / 2.0f + currentArrowMargin,
            (height - leftArrow.height) / 2.0f,
            indicatorPaint
        )

        // Draw the left and right icons
        canvas.drawBitmap(
            leftIcon,
            sidePadding,
            (height - leftIcon.height) / 2.0f,
            dataLinePaint
        )
        canvas.drawBitmap(
            rightIcon,
            width - rightIcon.width - sidePadding,
            (height - rightIcon.height) / 2.0f,
            dataLinePaint
        )

        // Draw the drag indicator button
        canvas.drawCircle(dragButtonPosition.x, height / 2.0f, buttonSelectorRadius, dataLinePaint)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = MeasureSpec.getSize(widthMeasureSpec)
        val height = MeasureSpec.getSize(heightMeasureSpec)
        dragButtonPosition.x = width / 2f
        dragButtonPosition.y = height / 2f
        invalidate()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                checkIfIsSwiping(event)
                setButtonPosition(event)
                isSwiping = true
            }
            MotionEvent.ACTION_MOVE -> setButtonPosition(event)
            MotionEvent.ACTION_UP -> {
                setButtonPosition(event)

                // Check if let go above one of the option icons
                when (getHoverOverSide()) {
                    HoveringOverSide.NONE -> {
                        resetButtonPos()
                        isHoldSwiping = false
                        isSwiping = false
                    }

                    // Selected the left icon
                    HoveringOverSide.LEFT -> {
                        onLeftSideSelectedListener?.invoke()
                        fadeOut()
                    }

                    // Selected the right option
                    HoveringOverSide.RIGHT -> {
                        onRightSideSelectedListener?.invoke()
                        fadeOut()
                    }
                }
            }
        }

        invalidate()
        return true
    }

    private fun getCurrentAnimatedAlpha(): Float {
        val pos = currentArrowMargin / indicatorMoveTo.toFloat()
        if (requiresSwipeMotion && !isHoldSwiping || !requiresSwipeMotion && !isSwiping) {
            return max(min(7.48164 * pos.toDouble().pow(3.0) - 15.36474 * pos.toDouble().pow(2.0) + 7.88309 * pos, 1.0), 0.0).toFloat()
        }
        return 0f
    }

    private fun fadeOut() {
        ValueAnimator.ofFloat(1f, 0f).apply {
            duration = 300
            interpolator = LinearInterpolator()
            addUpdateListener {
                val value = it.getAnimatedValue() as Float
                setAlpha(value)
                if (value == 0.0f) {
                    onFadedAwayListener?.invoke()
                }
            }
        }.start()
    }

    private fun checkIfIsSwiping(event: MotionEvent) {
        val eventPosition = PointF(event.x, event.y)
        val distance = dragButtonPosition - eventPosition
        isHoldSwiping = distance.length() <= buttonSelectorRadius
    }

    private fun resetButtonPos() {
        dragButtonPosition.x = width / 2.0f
        arrowAnimation.cancel()
        arrowAnimation.start()
    }

    private fun setButtonPosition(event: MotionEvent) {
        if (!requiresSwipeMotion || isHoldSwiping) {
            val xMin = sidePadding + leftIcon.width / 2f
            val xMax = width - sidePadding - rightIcon.width / 2f
            dragButtonPosition.x = min(max(event.x, xMin), xMax)
        }
    }

    private fun getHoverOverSide(): HoveringOverSide {
        if (dragButtonPosition.x <= sidePadding + leftIcon.width) {
            return HoveringOverSide.LEFT
        } else if (dragButtonPosition.x >= width - sidePadding - rightIcon.width) {
            return HoveringOverSide.RIGHT
        }
        return HoveringOverSide.NONE
    }

    internal enum class HoveringOverSide {
        NONE,
        LEFT,
        RIGHT
    }
}